typings/index.d.ts

```diff
@@ -278,23 +278,25 @@ export class Channel extends Base {
   public toString(): ChannelMention;
 }
 
-export class Client extends BaseClient {
+type If<T extends boolean, A, B = null> = T extends true ? A : T extends false ? B : A | B;
+
+export class Client<Ready extends boolean = boolean> extends BaseClient {
   public constructor(options: ClientOptions);
   private actions: unknown;
   private _eval(script: string): unknown;
   private _validateOptions(options: ClientOptions): void;
 
-  public application: ClientApplication | null;
+  public application: If<Ready, ClientApplication>;
   public channels: ChannelManager;
   public readonly emojis: BaseGuildEmojiManager;
   public guilds: GuildManager;
   public options: ClientOptions;
-  public readyAt: Date | null;
-  public readonly readyTimestamp: number | null;
-  public shard: ShardClientUtil | null;
-  public token: string | null;
-  public readonly uptime: number | null;
-  public user: ClientUser | null;
+  public readyAt: If<Ready, Date>;
+  public readonly readyTimestamp: If<Ready, number>;
+  public shard: If<Ready, ShardClientUtil>;
```

Comment:
Isn't this set in the constructor, and is independent on the client logging in?

Reply:
Ah right, will update